clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiFuture.java

```diff
@@ -109,7 +110,7 @@ private void completeExceptionally(K key, Throwable t) {
         }
 
         private KafkaFutureImpl<V> futureOrThrow(K key) {
-            KafkaFutureImpl<V> future = futures.get(key);
+            KafkaFutureImpl<V> future = (KafkaFutureImpl<V>) futures.get(key);
```

Comment:
I understand why we need to cast the `KafkaFuture` (parent) class into `KafkaFutureImpl` (child) class here since there are some methods only existed (or public) in child. But I think it'd better to add comment here to explain that this is a safe cast because the future is always `KafkaFutureImpl` instances. If this assumption changed, this should be updated. WDYT?

Reply:
The typecast isn't actually necessary if we're willing to declare `public Map<K, ? extends KafkaFuture<V>> all()` in `SimpleAdminApiFuture` and also in a handful of the `*Result` classes which use the `SimpleAdminApiFuture`. Those `*Result` classes have package access constructors so it would be compatible to do that. And that's actually the correct way to solve this. But I think there are other `Result` classes where that approach, if used more consistently in the other `*Result`, would be problematic. The precedent was set a long time ago to just use plain `Map<X, Y>` rather than the more correct `Map<X, ? extends Y>` in the Result classes. So I guess the typecast is the lesser evil, and I'm OK with a comment.